ethcore/src/verification/verification.rs

```diff
@@ -59,63 +60,66 @@ impl HeapSizeOf for PreverifiedBlock {
 }
 
 /// Phase 1 quick block verification. Only does checks that are cheap. Operates on a single block
-pub fn verify_block_basic(header: &Header, bytes: &[u8], engine: &EthEngine) -> Result<(), Error> {
-	verify_header_params(&header, engine, true)?;
-	verify_block_integrity(bytes, &header.transactions_root(), &header.uncles_hash())?;
-	engine.verify_block_basic(&header)?;
-	for u in Rlp::new(bytes).at(2)?.iter().map(|rlp| rlp.as_val::<Header>()) {
-		let u = u?;
-		verify_header_params(&u, engine, false)?;
-		engine.verify_block_basic(&u)?;
+pub fn verify_block_basic(block: &Unverified, engine: &EthEngine) -> Result<(), Error> {
+	verify_header_params(&block.header, engine, true)?;
+	verify_block_integrity(block)?;
+	engine.verify_block_basic(&block.header)?;
+
+	for uncle in &block.uncles {
+		verify_header_params(uncle, engine, false)?;
+		engine.verify_block_basic(uncle)?;
 	}
 
-	for t in Rlp::new(bytes).at(1)?.iter().map(|rlp| rlp.as_val::<UnverifiedTransaction>()) {
-		engine.verify_transaction_basic(&t?, &header)?;
+	for t in &block.transactions {
+		engine.verify_transaction_basic(t, &block.header)?;
 	}
+
 	Ok(())
 }
 
 /// Phase 2 verification. Perform costly checks such as transaction signatures and block nonce for ethash.
 /// Still operates on a individual block
 /// Returns a `PreverifiedBlock` structure populated with transactions
-pub fn verify_block_unordered(header: Header, bytes: Bytes, engine: &EthEngine, check_seal: bool) -> Result<PreverifiedBlock, Error> {
+pub fn verify_block_unordered(block: Unverified, engine: &EthEngine, check_seal: bool) -> Result<PreverifiedBlock, Error> {
+	let header = block.header;
 	if check_seal {
 		engine.verify_block_unordered(&header)?;
-		for u in Rlp::new(&bytes).at(2)?.iter().map(|rlp| rlp.as_val::<Header>()) {
-			engine.verify_block_unordered(&u?)?;
+		for uncle in &block.uncles {
+			engine.verify_block_unordered(uncle)?;
 		}
 	}
 	// Verify transactions.
-	let mut transactions = Vec::new();
 	let nonce_cap = if header.number() >= engine.params().dust_protection_transition {
 		Some((engine.params().nonce_cap_increment * header.number()).into())
-	} else { None };
-	{
-		let v = view!(BlockView, &bytes);
-		for t in v.transactions() {
+	} else {
+		None
+	};
+
+	let transactions = block.transactions
+		.into_iter()
+		.map(|t| {
 			let t = engine.verify_transaction_unordered(t, &header)?;
 			if let Some(max_nonce) = nonce_cap {
 				if t.nonce >= max_nonce {
 					return Err(BlockError::TooManyTransactions(t.sender()).into());
```

Comment:
Any reasons why we have `return Err(....into())` in some places and `bail!(...)` in other ones apart from historical?